garage/tf/regressors/categorical_mlp_regressor_with_model.py

```diff
@@ -0,0 +1,255 @@
+                 name='CategoricalMLPRegressorWithModel',
+                 hidden_sizes=(32, 32),
+                 hidden_nonlinearity=tf.nn.tanh,
+                 hidden_w_init=tf.glorot_uniform_initializer(),
+                 hidden_b_init=tf.zeros_initializer(),
+                 output_nonlinearity=tf.nn.softmax,
+                 output_w_init=tf.glorot_uniform_initializer(),
+                 output_b_init=tf.zeros_initializer(),
+                 optimizer=None,
+                 optimizer_args=None,
+                 tr_optimizer=None,
+                 tr_optimizer_args=None,
+                 use_trust_region=True,
+                 max_kl_step=0.01,
+                 normalize_inputs=True,
+                 layer_normalization=False,
+                 no_initial_trust_region=True):
+
+        super().__init__(input_shape, output_dim, name)
+        self._use_trust_region = use_trust_region
+        self._no_initial_trust_region = no_initial_trust_region
+        self._max_kl_step = max_kl_step
+        self._normalize_inputs = normalize_inputs
+        self._first_optimized = not no_initial_trust_region
+
+        with tf.variable_scope(
+                self._name, reuse=False) as self._variable_scope:
+            if optimizer is None:
+                optimizer = LbfgsOptimizer()
+            if tr_optimizer is None:
+                tr_optimizer = ConjugateGradientOptimizer()
+            self._optimizer = optimizer
+            self._tr_optimizer = tr_optimizer
+
+        self.model = NormalizedInputMLPModel(
+            input_shape,
+            output_dim,
+            hidden_sizes=hidden_sizes,
+            hidden_nonlinearity=hidden_nonlinearity,
+            hidden_w_init=hidden_w_init,
+            hidden_b_init=hidden_b_init,
+            output_nonlinearity=output_nonlinearity,
+            output_w_init=output_w_init,
+            output_b_init=output_b_init,
+            layer_normalization=layer_normalization)
+
+        self._initialize()
+
+    def _initialize(self):
+        input_var = tf.placeholder(
+            tf.float32, shape=(None, ) + self._input_shape)
+
+        with tf.variable_scope(self._variable_scope):
+            self.model.build(input_var)
+
+            ys_var = tf.placeholder(
+                dtype=tf.float32, name='ys', shape=(None, self._output_dim))
+
+            old_prob_var = tf.placeholder(
+                dtype=tf.float32,
+                name='old_prob',
+                shape=(None, self._output_dim))
+
+            y_hat = self.model.networks['default'].y_hat
+
+            old_info_vars = dict(prob=old_prob_var)
+            info_vars = dict(prob=y_hat)
+
+            self._dist = Categorical(self._output_dim)
+            mean_kl = tf.reduce_mean(
+                self._dist.kl_sym(old_info_vars, info_vars))
+
+            loss = -tf.reduce_mean(
+                self._dist.log_likelihood_sym(ys_var, info_vars))
+
+            predicted = tensor_utils.to_onehot_sym(
```

Comment:
isn't there a TF API for this?

Reply:
if so, please remove this (also remove it from tensor_utils and replace usages elsewhere with the TF API)

Reply:
There is a TF API, and this usage is just a wrapper to TF function. Will probably replace it with TF API here and separately address other usages in a different PR.

Reply:
lgtm